### ISSUE TITLE
JavaScript heap out of memory restoring 15 GB database

### ISSUE THREAD

**Issue by ricellis** [cloudant/couchbackup#134]
Please include the following information in your ticket:

- couchbackup/couchrestore version(s) that are affected by this issue.
>2.0.0-SNAPSHOT
- Node.js version.
>8.1.2
- A sample of the code or command line statement that demonstrates the issue.
>couchrestore --db abc123

When trying to restore the backup of a 15 GB database the node process runs out of memory. When using the API this restore process worked, but the backup file was also generated by the API in that case so may not have been identical.

```
[Node] Wed, 28 Jun 2017 15:56:56 GMT couchbackup:restore restored 3287500
[Node] 
[Node] <--- Last few GCs --->
[Node] 
[Node] [1344:0x39d7b30] 29523972 ms: Mark-sweep 17592186044402.6 (25.1) -> 17592186044402.5 (25.1) MB, 28.8 / 0.0 ms  last resort 
[Node] [1344:0x39d7b30] 29524006 ms: Mark-sweep 17592186044402.5 (25.1) -> 17592186044402.5 (25.1) MB, 33.7 / 0.0 ms  last resort 
[Node] 
[Node] 
[Node] <--- JS stacktrace --->
[Node] 
[Node] ==== JS stack trace =========================================
[Node] 
[Node] 
[Node] ==== Details ================================================
[Node] 
[Node] =====================
[Node] 
[Node] 
[Node] FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
[Node]  1: node::Abort() [/home/jenkins/.nvm/versions/node/v8.1.2/bin/node]
[Node]  2: 0x13647ec [/home/jenkins/.nvm/versions/node/v8.1.2/bin/node]
[Node]  3: v8::Utils::ReportOOMFailure(char const*, bool) [/home/jenkins/.nvm/versions/node/v8.1.2/bin/node]
[Node]  4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/home/jenkins/.nvm/versions/node/v8.1.2/bin/node]
[Node]  5: v8::internal::Factory::NewFixedArray(int, v8::internal::PretenureFlag) [/home/jenkins/.nvm/versions/node/v8.1.2/bin/node]
[Node]  6: v8::internal::HashTable<v8::internal::StringTable, v8::internal::StringTableShape, v8::internal::HashTableKey*>::New(v8::internal::Isolate*, int, v8::internal::PretenureFlag) [/home/jenkins/.nvm/versions/node/v8.1.2/bin/node]
[Node]  7: v8::internal::HashTable<v8::internal::StringTable, v8::internal::StringTableShape, v8::internal::HashTableKey*>::New(v8::internal::Isolate*, int, v8::internal::MinimumCapacity, v8::internal::PretenureFlag) [/home/jenkins/.nvm/versions/node/v8.1.2/bin/node]
[Node]  8: v8::internal::HashTable<v8::internal::StringTable, v8::internal::StringTableShape, v8::internal::HashTableKey*>::EnsureCapacity(v8::internal::Handle<v8::internal::StringTable>, int, v8::internal::HashTableKey*, v8::internal::PretenureFlag) [/home/jenkins/.nvm/versions/node/v8.1.2/bin/node]
[Node]  9: v8::internal::StringTable::LookupKey(v8::internal::Isolate*, v8::internal::HashTableKey*) [/home/jenkins/.nvm/versions/node/v8.1.2/bin/node]
[Node] 10: v8::internal::AstValueFactory::Internalize(v8::internal::Isolate*) [/home/jenkins/.nvm/versions/node/v8.1.2/bin/node]
[Node] 11: 0xd98e50 [/home/jenkins/.nvm/versions/node/v8.1.2/bin/node]
[Node] 12: 0xd9ce59 [/home/jenkins/.nvm/versions/node/v8.1.2/bin/node]
[Node] 13: v8::internal::Compiler::Compile(v8::internal::Handle<v8::internal::JSFunction>, v8::internal::Compiler::ClearExceptionFlag) [/home/jenkins/.nvm/versions/node/v8.1.2/bin/node]
[Node] 14: v8::internal::Runtime_CompileLazy(int, v8::internal::Object**, v8::internal::Isolate*) [/home/jenkins/.nvm/versions/node/v8.1.2/bin/node]
[Node] 15: 0x32e1fd0437d
[Node] Wed, 28 Jun 2017 15:56:57 GMT couchbackup:restore ERROR { BackupFileJsonError: Error on line 1647 of backup file - cannot parse as JSON
[Node]     at Transform.writer._transform (/home/jenkins/workspace/test-couchbackup-tests/includes/writer.js:155:26)
[Node]     at Transform._read (_stream_transform.js:190:10)
[Node]     at Transform._write (_stream_transform.js:178:12)
[Node]     at doWrite (_stream_writable.js:371:12)
[Node]     at writeOrBuffer (_stream_writable.js:357:5)
[Node]     at Transform.Writable.write (_stream_writable.js:274:11)
[Node]     at Transform.ondata (_stream_readable.js:628:20)
[Node]     at emitOne (events.js:115:13)
[Node]     at Transform.emit (events.js:210:7)
[Node]     at addChunk (_stream_readable.js:252:12) name: 'BackupFileJsonError' }
[Node] Aborted (core dumped)
```

**Comment by ricellis** [comment]
I was able to successfully restore a 15 GB database without error using the latest master. It is clear that in this failure case the backup file was also corrupted so that may be a factor. I suspect further testing is needed.

**Comment by ricellis** [comment]
As stated in the previous comment we've successfully backed up and restored 15 GB on master since this was opened. In that time we've also seen other 15 GB and larger backups and restores complete without issues. Further to that I've tested a corrupted 15 GB backup file and been able to restore it without the process running out of memory. Since I can't reproduce this - I think it has likely been fixed elsewhere or is related to the test framework running out of memory under some conditions rather than an issue with the `couchbackup` tool itself.
